simcon_project/conversation_templates/views/template_responses_table.py

```diff
@@ -0,0 +1,85 @@
+from django.contrib.auth.mixins import LoginRequiredMixin, UserPassesTestMixin
+from django.shortcuts import render, redirect
+from django.urls import reverse
+from django_tables2 import tables, RequestConfig, SingleTableView
+from django_tables2.export.views import TableExport
+from conversation_templates.models import ConversationTemplate, TemplateResponse
+from conversation_templates.forms import SelectTemplateForm
+
+
+def is_researcher(user):
+    return user.is_authenticated and user.is_researcher
+
+
+class ResponseTable(tables.Table):
+    first = tables.columns.Column()
+    last = tables.columns.Column()
+
+    class Meta:
+        model = TemplateResponse
+        fields = ['first', 'last', 'completion_date']
+
+
+class TemplateResponsesView(UserPassesTestMixin, LoginRequiredMixin, SingleTableView):
+    model = TemplateResponse
+    table_class = ResponseTable
+    template_name = "template_all_responses_view.html"
+
+    def test_func(self):
+        return self.request.user.is_authenticated and self.request.user.is_researcher
+
+    def get(self, request, pk):
+        """
+        On get request render a custom table based on ResponseTable:
+        Rows: all template responses for the conversation template
+        Columns: student.first_name, student.last_name, template_response.completion_date,
+                 and template_node_responses with template_node.description as headers (this is
+                 dynamically created since each template has different number of nodes")
+
+        TODO: make sure incomplete responses aren't shown
+        """
+        template = ConversationTemplate.objects.get(pk=pk)
+        descriptions = []  # Description of template node used for column header
+        extra_columns = []  # List of tuples of description and column object to pass to table
+        table_data = []  # List of dictionaries to populate table. 1 dictionary = 1 column
+
+        for node in template.template_nodes.all():
+            descriptions.append(node.description)
+            extra_columns.append((node.description, tables.columns.Column()))
+
+        # Populate the table with data for each response
+        for response in template.template_responses.all().order_by('-completion_date'):
+            column_data = {
+                "first": response.student.first_name,
+                "last": response.student.last_name,
+                "completion_date": response.completion_date,
+            }
+
+            for idx, node in enumerate(response.node_responses.all()):
+                column_data.update({descriptions[idx]: node.transcription})
```

Comment:
I think there needs to be some check here to compare the template_node field in each node_response with the templates_nodes grabbed on lines 46-48. If they match then update with the node.transcription, otherwise update with an empty string or None. Probably will need another for loop methinks.

Reply:
I took the descriptions array out and just made it so I get the current template_node_responses parent template_node's description directly. Now it will match correctly to it's parent node in the table without another for loop!